client/client.go

```diff
@@ -450,52 +450,55 @@ func (c *client) processTSORequests(stream pdpb.PD_TsoClient, dcLocation string,
 
 	if err := stream.Send(req); err != nil {
 		err = errors.WithStack(err)
-		c.finishTSORequest(requests, 0, 0, err)
+		c.finishTSORequest(requests, 0, 0, 0, err)
 		return err
 	}
 	resp, err := stream.Recv()
 	if err != nil {
 		err = errors.WithStack(err)
-		c.finishTSORequest(requests, 0, 0, err)
+		c.finishTSORequest(requests, 0, 0, 0, err)
 		return err
 	}
 	requestDurationTSO.Observe(time.Since(start).Seconds())
 	tsoBatchSize.Observe(float64(count))
 
 	if resp.GetCount() != uint32(len(requests)) {
 		err = errors.WithStack(errTSOLength)
-		c.finishTSORequest(requests, 0, 0, err)
+		c.finishTSORequest(requests, 0, 0, 0, err)
 		return err
 	}
 
-	physical, logical := resp.GetTimestamp().GetPhysical(), resp.GetTimestamp().GetLogical()
+	physical, logical, suffixBits := resp.GetTimestamp().GetPhysical(), resp.GetTimestamp().GetLogical(), resp.GetTimestamp().GetSuffixBits()
 	// Server returns the highest ts.
-	logical -= int64(resp.GetCount() - 1)
-	c.compareAndSwapTS(dcLocation, physical, logical, int64(len(requests)))
-	c.finishTSORequest(requests, physical, logical, nil)
+	firstLogical := logical - int64((count-1)<<suffixBits)
+	c.compareAndSwapTS(dcLocation, physical, firstLogical, int64(len(requests)))
+	c.finishTSORequest(requests, physical, firstLogical, suffixBits, nil)
 	return nil
 }
 
-func (c *client) compareAndSwapTS(dcLocation string, physical, logical, n int64) {
+func (c *client) compareAndSwapTS(dcLocation string, physical, firstLogical, count int64) {
 	lastTSOInterface, ok := c.lastTSMap.Load(dcLocation)
 	if !ok {
 		var loaded bool
 		if lastTSOInterface, loaded = c.lastTSMap.LoadOrStore(dcLocation, &lastTSO{
 			physical: physical,
-			logical:  logical + n - 1,
+			// Save the highest logical part here
```

Comment:
one question, why not use `LoadOrStore` directly on L480, so L480-L490 can reduce to:
```
lastTSOInterface, loaded := c.lastTSMap.LoadOrStore(dcLocation, &lastTSO{
		physical: physical,
		// Save the highest logical part here
		logical: firstLogical + count - 1,
	})
	if !loaded {
		return
	}
```
which I think is easier to understand?

Reply:
You're right. `LoadOrStore` can already guarantee atomicity. I misunderstand it.